compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -1473,22 +1474,32 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                         let ty_from = op.ty(self.body, tcx);
                         let cast_ty_from = CastTy::from_ty(ty_from);
                         let cast_ty_to = CastTy::from_ty(*ty);
+
                         match (cast_ty_from, cast_ty_to) {
                             (Some(CastTy::Ptr(src)), Some(CastTy::Ptr(dst))) => {
-                                let src_tail = self.struct_tail(src.ty, location);
-                                let dst_tail = self.struct_tail(dst.ty, location);
-
-                                // This checks (lifetime part of) vtable validity for pointer casts,
-                                // which is irrelevant when there are aren't principal traits on
-                                // both sides (aka only auto traits).
-                                //
-                                // Note that other checks (such as denying `dyn Send` -> `dyn
-                                // Debug`) are in `rustc_hir_typeck`.
-                                if let ty::Dynamic(src_tty, _src_lt, ty::Dyn) = *src_tail.kind()
-                                    && let ty::Dynamic(dst_tty, dst_lt, ty::Dyn) = *dst_tail.kind()
+                                if self
+                                    .infcx
+                                    .type_is_sized_modulo_regions(self.infcx.param_env, dst.ty)
+                                {
+                                    // Wide to thin ptr cast. This may even occur in an env with
+                                    // impossible predicates, such as `where dyn Trait: Sized`.
+                                    // In this case, we don't want to fall into the case below,
+                                    // since the types may not actually be equatable, but it's
+                                    // fine to perform this operation in an impossible env.
```

Comment:
I don't really care about enforcing that this `Sized` bound holds considering regions.

Reply:
could slap a `prove_predicate` in this branch to check that, can't we?

I don't think it matters for soundness or anything but it feels weird to not do so here.